tests/integration_tests/tests/utils.py

```diff
@@ -200,3 +204,110 @@ def patch_yaml(yaml_path, is_json=False, default_flow_style=True):
 
 def delete_provider_context():
     postgresql.run_query('DELETE from provider_context')
+
+
+def delete_snapshot_if_existing(snapshot_id):
+    client = create_rest_client()
+
+    try:
+        client.snapshots.delete(snapshot_id)
+    except CloudifyClientError as err:
+        if 'not found' in str(err):
+            pass
+        else:
+            raise
+
+
+def upload_snapshot(path_to_snapshot, snapshot_id):
+    client = create_rest_client()
+
+    client.snapshots.upload(path_to_snapshot, snapshot_id)
+
+
+def restore_snapshot(snapshot_id, tenant):
+    client = create_rest_client()
+
+    return client.snapshots.restore(
+        snapshot_id,
+        tenant_name=tenant,
+        force=True,
+    )
+
+
+def upload_and_restore_snapshot(snapshot_file_name, snapshot_id, tenant):
+    path_to_snapshot = get_snapshot_from_s3(snapshot_file_name)
+
+    # This and the force setting allow multiple tests to run on the same
+    # container (the default for agentless tests)
+    delete_snapshot_if_existing(snapshot_id)
+
+    upload_snapshot(path_to_snapshot, snapshot_id)
+
+    validate_snapshot(snapshot_id, path_to_snapshot)
+
+    execution = restore_snapshot(snapshot_id, tenant)
+
+    # If checking the workflow status immediately, we will be likely to cause
+    # a deadlock on the database and make the restore fail
+    # This may be possible to do using the cfy utility, but I've not managed
+    # to be fast enough with it to trigger the issue.
+    # NOTE: Jira should be opened for this
+    time.sleep(3)
```

Comment:
Well, you can write your own function and use the `retrying.retry` decorator.
It can be retried for any error until the execution's status == terminated.